bilby/gw/likelihood.py

```diff
@@ -1616,35 +1616,38 @@ def compute_summary_data(self):
-            factor = masked_frequency_array - maximum_bin_frequency_array
-
-            a0 = noise_weighted_inner_product(
-                interferometer.frequency_domain_strain[mask],
-                self.per_detector_fiducial_waveforms[interferometer.name][mask],
-                interferometer.power_spectral_density_array[mask],
-                self.waveform_generator.duration) / self.number_of_bins
-            b0 = noise_weighted_inner_product(
-                self.per_detector_fiducial_waveforms[interferometer.name][mask],
-                self.per_detector_fiducial_waveforms[interferometer.name][mask],
-                interferometer.power_spectral_density_array[mask],
-                self.waveform_generator.duration) / self.number_of_bins
-            a1 = noise_weighted_inner_product(
-                interferometer.frequency_domain_strain[mask] * factor,
-                self.per_detector_fiducial_waveforms[interferometer.name][mask],
-                interferometer.power_spectral_density_array[mask],
-                self.waveform_generator.duration) / self.number_of_bins
-            b1 = noise_weighted_inner_product(
-                self.per_detector_fiducial_waveforms[interferometer.name][mask] * factor,
-                self.per_detector_fiducial_waveforms[interferometer.name][mask],
-                interferometer.power_spectral_density_array[mask],
-                self.waveform_generator.duration) / self.number_of_bins
+                masked_bin_inds.append(index)
+            masked_strain = interferometer.frequency_domain_strain[mask]
+            masked_h0 = self.per_detector_fiducial_waveforms[interferometer.name][mask]
+            masked_psd = interferometer.power_spectral_density_array[mask]
+
+            a0 = np.array([noise_weighted_inner_product(
+                masked_strain[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_h0[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_psd[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                self.waveform_generator.duration) for i in range(self.number_of_bins)])
+            b0 = np.array([noise_weighted_inner_product(
+                masked_h0[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_h0[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_psd[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                self.waveform_generator.duration) for i in range(self.number_of_bins)])
+            a1 = np.array([noise_weighted_inner_product(
+                masked_strain[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_h0[masked_bin_inds[i]:masked_bin_inds[i + 1]],
+                masked_psd[masked_bin_inds[i]:masked_bin_inds[i + 1]]
+                * masked_frequency_array[masked_bin_inds[i]:masked_bin_inds[i + 1]]
+                - 0.5 * (masked_frequency_array[masked_bin_inds[i]] + masked_frequency_array[masked_bin_inds[i + 1]]),
```

Comment:
Just for documentation purposes, could you describe what you have written `a1` in this form? Specifically, could you explain (mathematically) why the third argument (supposed to be the PSD) is changed to:

```
masked_psd[masked_bin_inds[i]:masked_bin_inds[i + 1]]
* masked_frequency_array[masked_bin_inds[i]:masked_bin_inds[i + 1]]
- 0.5 * (masked_frequency_array[masked_bin_inds[i]] + masked_frequency_array[masked_bin_inds[i + 1]])
```

Reply:
I think it's a mistake. It's the `f - f_m` factor that is supposed to be multiplied to either the first or second argument. But I don't know how the evaluations are matching then.

Reply:
There's also a factor of `duration**2` that I'm confused about. Thanks for pointing this out. I'll look into this more closely tomorrow.

Reply:
> I think it's a mistake. It's the f - f_m factor that is supposed to be multiplied to either the first or second argument. But I don't know how the evaluations are matching then.

If you are evaluating the likelihood for the fiducial parameters, then I guess it will match since `waveform_ratio` is 1 and `a1` or `b1` do not contribute to the calculation.

> There's also a factor of duration**2 that I'm confused about. Thanks for pointing this out. I'll look into this more closely tomorrow.

Where do you see this?

Reply:
Also, instead of using list comprehension for each of `a0, a1, b0, b1`, I think it would be better to just make a loop. You can initialize them to be zeros outside the loop and then just assign the elements.

Reply:
> If you are evaluating the likelihood for the fiducial parameters, then I guess it will match since waveform_ratio is 1 and a1 or b1 do not contribute to the calculation.

Yeah, but right after shifting the frequency_factor to second arguments, the likelihood evaluations were off by `duration**2 * 10^18`or so. That's pretty weird.

Reply:
Umm I am not sure I understand well, but could you repeat the same thing with the current version?

Reply:
I don't why I was getting that. But the current version is working properly!